crates/router/src/workflows/revenue_recovery.rs

```diff
@@ -485,8 +518,250 @@ impl From<InternalDeciderRequest> for external_grpc_client::DeciderRequest {
+        state,
+        &payment_attempt,
+        payment_intent,
+        token_with_retry_info,
+    )
+    .await
+    .map(|schedule_time| ScheduledToken {
+        token_details: token_status.payment_processor_token_details.clone(),
+        schedule_time,
+    });
+
+    Ok(scheduled_token)
+}
+
+#[cfg(feature = "v2")]
+pub async fn call_decider_for_payment_processor_tokens_select_closet_time(
+    state: &SessionState,
+    merchant_context: domain::MerchantContext,
+    processor_tokens: &HashMap<String, PaymentProcessorTokenWithRetryInfo>,
+    payment_id: &id_type::GlobalPaymentId,
+    connector_customer_id: &str,
+) -> Result<Option<ScheduledToken>, errors::ProcessTrackerError> {
+    let db = &*state.store;
+    let key_manager_state = &(state).into();
+    let storage_scheme = merchant_context.get_merchant_account().storage_scheme;
+
+    let payment_intent = db
+        .find_payment_intent_by_id(
+            key_manager_state,
+            payment_id,
+            merchant_context.get_merchant_key_store(),
+            storage_scheme,
+        )
+        .await?;
+
+    let mut scheduled_tokens: Vec<ScheduledToken> = Vec::new();
+
+    for (_token_id, token_with_retry_info) in processor_tokens.iter() {
+        let token_details = &token_with_retry_info
+            .token_status
+            .payment_processor_token_details;
+        let error_code = token_with_retry_info.token_status.error_code.clone();
+
+        match error_code {
+            None => {
+                let utc_schedule_time =
+                    time::OffsetDateTime::now_utc() + time::Duration::minutes(5);
+                let schedule_time = time::PrimitiveDateTime::new(
+                    utc_schedule_time.date(),
+                    utc_schedule_time.time(),
+                );
+
+                return Ok(Some(ScheduledToken {
+                    token_details: token_details.clone(),
+                    schedule_time,
+                }));
+            }
+            Some(_) => {
+                if let Some(scheduled_token) = process_token_for_retry(
+                    state,
+                    token_with_retry_info,
+                    merchant_context.clone(),
+                    &payment_intent,
+                )
+                .await?
+                {
+                    scheduled_tokens.push(scheduled_token);
+                }
+            }
+        }
+    }
+
+    let best_token = scheduled_tokens
+        .iter()
+        .min_by_key(|token| token.schedule_time)
+        .cloned();
+
+    match best_token {
+        Some(token) => {
+            RedisTokenManager::update_payment_processor_token_schedule_time(
+                state,
+                connector_customer_id,
+                &token.token_details.payment_processor_token,
+                Some(token.schedule_time),
+            )
+            .await?;
+            Ok(Some(token))
+        }
+        None => Ok(None),
+    }
```

Comment:
```suggestion
best_token
    .async_map(|token| async move {
        RedisTokenManager::update_payment_processor_token_schedule_time(
            state,
            connector_customer_id,
            &token.token_details.payment_processor_token,
            Some(token.schedule_time),
        )
        .await?;
        Ok(token)
    })
    .await
    .transpose()
```

 can we use asynv_map instead ?